### ISSUE TITLE
CLI: Specifying an external config causes ESLint to ignore per-directory overrides

### ISSUE THREAD

**Issue by fearphage** [eslint/eslint#13871]
When I specify a config file that is not in the current subdirectory tree, the local directory overrides are ignored.

## Steps to Reproduce

1. Create a basic config `extends: eslint:recommended` in a directory - `/tmp/foo` for instance
1. Go to a new directory that is **not** a subdirect of the one created in step 1 - `/tmp/bar`
1. Place some erroneous code - example: `var a = 1`
1. Create a subdirectory of the one current directory (from step 2) - `/tmp/bar/under-test`
1. Put some bad js in the newly created subdirectory: `var b = 2`
1. Add a local eslint config to the sub directory - example:
    ````
     rules:
       no-unsude-vars: off
    ```
1. Install `eslint@latest`
1. Run eslint using the config created in step 1 against the files in the current directory 
    ```shell
    eslint --config /tmp/foo/.eslintrc.yml .`
    ```

## Expected Result

The error in the subdirectory (`under-test`) should be ignored based on the local config.

## Actual Result

The local config is ignored/not parsed so ESLint shows the error in the subdirectory... even though the config attempts to disable that error.

## Script that reproduces the issue (bash)

```
#!/usr/bin/env bash

set -eu

cd $(mktemp -d /tmp/meh.XXXXXXX)

# setup base config
echo 'extends: eslint:recommended' > .eslintrc.yml

eslint_config_path=$(pwd)/.eslintrc.yml

cd $(mktemp -d /tmp/meh.XXXXXXX)

npm install eslint@latest &> /dev/null

echo 'var a = 1' > bad.js

# make sure it works at all
$(npm bin)/eslint --config $eslint_config_path . | grep --quiet bad.js

mkdir deeper

echo 'var b = 2' > deeper/catpants.js

# make sure it detects nested error
$(npm bin)/eslint --config $eslint_config_path . | grep --quiet catpants.js

# add directory based config
echo 'rules: { no-unused-vars: off }' > deeper/.eslintrc.yml

# The nested error should be ignored now
#
#     My expectation here is that the catpants.js file is not reported as an 
#     error since the rule is disabled. That is not the case.
$(npm bin)/eslint --config $eslint_config_path . | grep --quiet catpants.js \
  && echo 'failed' || echo 'success'
```

<details>
<summary>Environment Information</summary>

<!-- Paste your configuration below: -->
```
Environment Info:

Node version: v14.13.1
npm version: v6.14.8
Local ESLint version: Not found
Global ESLint version: Not found
```

</details>

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
extends: eslint:recommended
```

</details>

<!-- Paste the command you used to run ESLint: -->
```bash
eslint --config /path/to/config/eslintrc.yml .
```
## Notes

I see in the debug output that ESLint sees the file, but it doesn't parse it as config:

```shell
  eslint:file-enumerator Enter the directory: /tmp/meh.rIepT3z/deeper +18ms           
  eslint:file-enumerator Didn't match: .eslintrc.yml +0ms                     
```

According to the documentation, local configuration should be combined with command line arguments by default. Am I missing something?

**Comment by mdjermanovic** [comment]
Hi @fearphage, thanks for the issue!

I believe this works as intended, per the [--config specification](https://eslint.org/docs/user-guide/command-line-interface#-c-config):

> Options from this configuration file have precedence over the options from .eslintrc.* and package.json files.

There is a test case to confirm that behavior [here](https://github.com/eslint/eslintrc/blob/main/tests/lib/cascading-config-array-factory.js#L1063) (`specificConfigPath` is the path from `--config`, fixtures are [here](https://github.com/eslint/eslintrc/blob/main/tests/fixtures/config-hierarchy/file-structure.json)).

> I see in the debug output that ESLint sees the file, but it doesn't parse it as config:
> 
> ```shell
>   eslint:file-enumerator Enter the directory: /tmp/meh.rIepT3z/deeper +18ms           
>   eslint:file-enumerator Didn't match: .eslintrc.yml +0ms        
> ```

That's the info for files that should be linted. For configs, you should see something like:

```
eslintrc:config-array-factory Loading YAML config file: /tmp/meh.rIepT3z/deeper/.eslintrc.yml
eslintrc:config-array-factory Config file found: /tmp/meh.rIepT3z/deeper/.eslintrc.yml
```

(we had some debug info missing in the previous version, it should be there again with the latest ESLint v7.14.0)

**Comment by fearphage** [comment]
> I believe this works as intended, per the [--config specification](https://eslint.org/docs/user-guide/command-line-interface#-c-config):

That statement makes the previous line seem misleading then:

> If .eslintrc.* and/or package.json files are also used for configuration (i.e., --no-eslintrc was not specified), **the configurations will be merged**.

Emphasis added by me. Is it possible to re-enable per-directory overrides when using a config or alternatively can I set a per-directory config from the CLI somehow?

**Comment by mdjermanovic** [comment]
Configurations are merged, but the order is important. `--config` having precedence over `.eslintrc.*` means that it will be merged over them and thus may overwrite some settings:

project dir config <- subdir config <- `--config` config

In your example, `--config` config has `extends: eslint:recommended` which has `no-unused-vars: error`, so it overwrites `no-unused-vars: off` from the subdir config.

**Comment by mdjermanovic** [comment]
> Is it possible to re-enable per-directory overrides when using a config or alternatively can I set a per-directory config from the CLI somehow?

`--config` can be used as the only config, or to enforce additional settings over those defined in the project's `.eslintrc.*` files.

If you want to use a shared config as a base configuration that can be overridden, then you can specify that config in `extends` in `.eslintrc.yml`.
